src/networking.c

```diff
@@ -3362,19 +3363,42 @@ void readToQueryBuf(client *c) {
     }
 }
 
+#define REPL_MAX_READS_PER_IO_EVENT 25
+/** Keeps replica reading from the primary if recvq has data. */
+static bool shouldRepeatReadFromPrimary(client *c, int iteration) {
+    // If the client is not a primary replica, is closing, or flow control is disabled, no more reads.
+    if (!(c->flag.primary) || c->flag.close_asap) {
+        return 0;
+    }
+
+    if (iteration < REPL_MAX_READS_PER_IO_EVENT &&
+        c->is_qb_full_read) {
+        return 1;
+    }
+
+    return 0;
+}
+
+
 void readQueryFromClient(connection *conn) {
     client *c = connGetPrivateData(conn);
     /* Check if we can send the client to be handled by the IO-thread */
     if (postponeClientRead(c)) return;
 
     if (c->io_write_state != CLIENT_IDLE || c->io_read_state != CLIENT_IDLE) return;
 
-    readToQueryBuf(c);
+    bool shouldRepeat = false;
+    int iter = 0;
+    do {
+        readToQueryBuf(c);
 
-    if (handleReadResult(c) == C_OK) {
-        if (processInputBuffer(c) == C_ERR) return;
-    }
-    beforeNextClient(c);
+        if (handleReadResult(c) == C_OK) {
+            if (processInputBuffer(c) == C_ERR) return;
+        }
+        iter++;
+        shouldRepeat = shouldRepeatReadFromPrimary(c, iter);
+        beforeNextClient(c);
+    } while (shouldRepeat);
```

Comment:
Does shouldRepeatReadFromPrimary need to be called before beforeNextClient? If not, we can do the check in the while condition:

```suggestion
        beforeNextClient(c);
    } while (shouldRepeatReadFromPrimary(c, iter));
```

Reply:
Yes. `shouldRepeatReadFromPrimary` must run before `beforeNextClient`, since the latter may free the client.